### ISSUE TITLE
owncloudcmd with Open ID Connect

### ISSUE THREAD

**Issue by dragotin** [owncloud/client#10256]
### Pre-submission Checks

- [X] I [checked for similar issues](https://github.com/owncloud/client/issues?q=), but could not find any. I also checked the closed issues. I could not contribute additional information to any existing issue.
- [X] I will take the time to fill in all the required fields. I know that the bug report may be dismissed otherwise due to lack of information.

### Feature description

`owncloudcmd` should be able to authenticate via OpenID Connect to run with oCIS.

### Proposed solution

Use the existing code in the client to re-use that with the commandline client.

### Alternative solutions you considered

It was considered to use the [`oidc-agent`](https://owncloud.dev/clients/rclone/webdav-sync-oidc/) that works with for example rclone - but since we have the code already in the client it seems easier and more convenient to use that in owncloudcmd as well.

**Comment by michaelstingl** [comment]
> Use the existing code in the client to re-use that with the commandline client.

How do you imagine the user experience? Open the GUI browser on the same machine? I'd love a solution that would allow to perform the browser operations on a different device, to allow all kind of headless scenarios (IoT, CI/CD), maybe based on the [`OAuth 2.0 Device Authorization Grant`](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-device-flow-15). The spec is still draft, but it has already been implemented by many major IdPs. (Keycloak 13+)

**Comment by TheOneRing** [comment]
Turns out to require bigger refactoring. We have several abstraction layers that are all pointless by now.
But we have a pretty implementation in ui.
We could  refactor the ui part out of it and inherit it in a cmd and gui class.

**Comment by dragotin** [comment]
Hm, any chance we can build something around `oidc-agent`, as rclone does it: https://rclone.org/webdav/#openid-connect ?

**Comment by TheOneRing** [comment]
Its a C lib that we need to build on 3+ platforms...

**Comment by fmoc** [comment]
I don't think that `oidc-agent` is a viable option. I could imagine that given the current state of our credentials related code base, reworking the GUI code to work with some CLI browser would take a similar amount of time. Plus it's plain C and adds additional dependencies.

In any case, we need to decide whether to use some technically clean solution (like RFC 8628 mentioned by @michaelstingl) or some workaround that tries to recycle the workflow that works well on a desktop for a commandline application. Please bear in mind use cases on headless systems like servers or CIs, where workflows like "open a browser to log in" just don't work well or don't work at all.

I'd like to play a bit with RFC 8628 and Keycloak to see whether that would be feasible. The advantage is that you can easily decouple the system you want to authorize from the system where you set up the token. This makes generating a token relatively easy for users. This `oidc-agent` project seems drastically more complicated to use.

**Comment by dragotin** [comment]
Thanks. Have you guys looked into the link I sent? I talked about a command line tool that generates/manages OIDC tokens: https://indigo-dc.gitbook.io/oidc-agent/

**Comment by fmoc** [comment]
Yes, I had a look by now. It could be used without the agent, as far as I can tell. We don't need to use it. But it's some external dependency.

What's good is we can use it on any host and then forward the token to `owncloudcmd`.

**Comment by fmoc** [comment]
We still don't have a solution to using the cmd client with oCIS and OIDC, therefore reopening. Maybe it's just a docs issue, but even then I think we could improve the user experience and make using it with oCIS easier and more straightforward.

**Comment by nicholas-wilson-au** [comment]
It would be good to add a specific Client ID and Client secret to oCIS for OIDC-Agent rather than reusing the Desktop app credentials:
client-id=xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69
client-secret=UBntmLjC2yYCeHwsyj73Uwo9TAaecAetRwMw0xYcvNL9yRdLSUi0hUAHfvCHFeFh

This would provide better distinction to the user, for debugging and for logging.

**Comment by fmoc** [comment]
Let's see what we can do in 6.0 about it. The change is likely going to require quite some refactoring in the OAuth2/OIDC code, but we need to clean that up anyway at some point, so why not in the next release?

**Comment by michaelstingl** [comment]
> It would be good to add a specific Client ID and Client secret to oCIS for OIDC-Agent rather than reusing the Desktop app credentials:

@nicholas-wilson-au issue in oCIS repo please

**Comment by dj4oC** [comment]
May I ask a status here?

**Comment by michaelstingl** [comment]
> May I ask a status here?

Candidate for the 6.0 release around X-mas, but there are more priority items probably.

**Comment by michaelstingl** [comment]
> May I ask a status here?

@dj4oC Work planned for next sprint, starting Nov. 20th, to be demoed in the sprint reviews Friday, Dec. 8th, then planned to be released in the 6.0 after X-mas…

**Comment by michaelstingl** [comment]
Would be nice to have WebDAV URL's visible in the web UI then…
- https://github.com/owncloud/web/issues/9714

/cc @tbsbdr @kulmann @JammingBen als for users, not only admins

**Comment by fmoc** [comment]
We plan to build a prototype setup until next sprint using oidc-agent. We are likely going to use subprocesses rather than the C library provided by the tool as a starting point.

The user will have to use oidc-agent to log into the specified IdP (both internal and external should be supported) and provide a "profile" within oidc-agent's secrets management. When calling `owncloudcmd`, said profile must be specified instead of username/password.

To request the initial tokens, a command similar to the following may be used:

```sh
oidc-gen <profile> \
    --iss https://keycloak.ocis-keycloak.latest.owncloud.works \
    --scope "openid email profile offline_access" \
    --client-id "xdXOt13JKxym1B1QcEncf2XDkLAexMBFwiT9j6EfhhHFJhs2KM9jbjTmf8JBXE69" \
    --client-secret "UBntmLjC2yYCeHwsyj73Uwo9TAaecAetRwMw0xYcvNL9yRdLSUi0hUAHfvCHFeFh" \
    --redirect-uri "http://localhost:9999"
```

Keycloak client ID and secret can be found in https://github.com/owncloud/ocis/blob/master/deployments/examples/ocis_keycloak/config/keycloak/clients/desktop_client.json. The redirect URI is used by oidc-agent to spawn an HTTP server to receive the IdP response (there is an alternative workflow available, though, that does not depend on that); it must be a value that is allowed by the IdP. For ownCloud's native clients, localhost redirects are used already, so this should work as-is.

Once the tokens have been set up, we are going to use `oidc-token <profile>` internally to request access tokens. When they expire, we can just request a new one. We need to have some retry mechanism so that if the tokens fail repeatedly (or `oidc-token` fails), we will abort the sync run.

In future versions, we will be able to provide a better experience, e.g., by automating requesting the tokens.
